feat: deleted data should not be retained

beliayeu opened the pull request:
<!--
  Before issuing a pull request:

  - Please read our contribution guidelines (https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/packages/bodiless-documentation/doc/Development/Contributing.md), especially the section on Pull Requests.
  - Please first create an issue (https://github.com/johnsonandjohnson/Bodiless-JS/issues/new). All pull requests must be linked to an issue.
-->

<!--
  IMPORTANT
  
  The pull request title will become the commit message title at merge, and
  should adhere to Angular Commit Message Conventions. Please see (https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/packages/bodiless-documentation/doc/Development/Contributing.md) for details and examples.
-->

## Changes
<!-- Brief description of the changes introduced by this PR -->

## Test Instructions
<!-- Detailed description of how this feature was (and should be) tested
  - Setup instructions (if any)?
  - List of test cases, with steps and expected results?
  - List
-->

## Related Issues
<!--
  Link to the issue that is fixed or resolved by this PR (if there is one)
  e.g. Fixes #1234, Closes #4567

  Link to an issue that is partially addressed by this PR (if there are any)
-->

<!-- IF THIS PR INTRODUCES A BREAKING CHANGE

BREAKING CHANGE: Describe the nature of the breaking change here.

More Details about the breaking change.
-->

beliayeu committed:
feat: deleted data should not be retained

wodenx commented on packages/gatsby-theme-bodiless/src/dist/GatsbyMobxStore.ts:
```diff
@@ -290,4 +150,25 @@ export default class GatsbyMobxStore {
+
+  deleteNode = (keyPath: string[]) => {
+    const children = this.getChildrenNodes(keyPath);
+    children.forEach(child => {
```
Would it be better to implement this logic on the backend, rather than sending a slew or delete requests?  I know ti's less RESTful - but should we be concerned about a large number of near-simultaneous http requests?

beliayeu replied:
@wodenx, the concern is whether we have a large number delete requests. we have delete requests triggered by list component and by flexbox. In the list case, we can get a lot of requests when we delete an item that has a lot of children. not sure it will happen very often.
in the flexbox case, we can get a lot of requests when we delete an item with a lot of children nodes.  not sure it will happen very often either.
if amount of requests becomes a problem for us, then I believe we can consider consolidating the requests on backend client side and/or change http to websockets

wodenx commented on packages/gatsby-theme-bodiless/src/dist/GatsbyMobxStore.ts:
```diff
@@ -290,4 +150,25 @@ export default class GatsbyMobxStore {
+  getChildrenNodes = (keyPath: string[]) => {
+    const key = keyPath.join(nodeChildDelimiter);
+    const children = Array.from(this.store)
+      .filter(item => item[0].indexOf(key) === 0 && item[0] !== key);
```
Why do we bother excluding the key here when we just have to make a separate call to delete the parent later?

beliayeu replied:
@wodenx, my idea was to remove all node children first, then the node itself, hereby I implemented a function to get a list of children. but I agree it is not important in our case

wodenx commented on packages/gatsby-theme-bodiless/src/dist/GatsbyMobxStoreItem.ts:
```diff
@@ -0,0 +1,179 @@
+
+  @observable state: ItemState = ItemState.Clean;
+
+  isDeleted = false;
```
Shouldn't this also be observable? We need a way to notify observers that the item has been deleted. It's probably not an issue now, bc we only delete items which don't have an active reference, but later we may want to delete items which are still visible (eg to revert to default content).

One option here is to create a @computed data getter method which respects `isDeleted` and returns empty data if the item has been deleted.

beliayeu replied:
@wodenx, it makes sense to make isDeleted as observable, but tbh I would defer it until the case when it is needed unless we have a use case know in which we can test it

beliayeu committed:
Merge branch 'fix/content-lost' into feat/retained-data

beliayeu committed:
Merge branch 'fix/content-lost' into feat/retained-data

beliayeu committed:
Merge branch 'fix/content-lost' into feat/retained-data

beliayeu committed:
Merge branch 'fix/content-lost' into feat/retained-data

beliayeu committed:
fixed build errors

beliayeu committed:
when item is marked as deleted then getnode returns empty

beliayeu committed:
updated flexbox mock

beliayeu committed:
Merge branch 'fix/content-lost' into feat/retained-data

beliayeu committed:
added unit tests

beliayeu committed:
added unit tests

dewen committed:
feat(layout): Easily identify a flexbox area on page so that can start… (#99)

* feat(layout): Easily identify a flexbox area on page so that can start adding components.

* feat(layout): Easily identify a flexbox area on page so that can start adding components.

* feat(layout): Easily identify a flexbox area on page so that can start adding components. Fixing TS error.

* feat(layout): Easily identify a flexbox area on page so that can start adding components. Border styling on select/hover.

Co-authored-by: hvanyo <58534559+hvanyo@users.noreply.github.com>

TatsianaMiraniuk committed:
test(organisms): Add smoke test for touts (#136)

## Changes
- Added smoke test for Touts
- Added a package to support the Xpath selectors
- Added a package to support uploading a file
- Set the default browser viewport sizes
- Added two custom commands

## Test Instructions
 - NOTE: due to a content loss issue (#14) the test may fail in Master

beliayeu committed:
removed list and flexbox json files that contains empty array

beliayeu committed:
fixed failed tests

konst8 committed:
feat(layout): Create a location Switcher for the Edit Admin UI Menu (#149)

dewen committed:
fix(migration_tool): issues in FAQ pages (#126)

* Migration tool: issues in FAQ pages

* Migration tool: issues in FAQ pages

* Migration tool: issues in FAQ pages

* Migration tool: issues in FAQ pages

* Migration tool: issues in FAQ pages - revert changes in starter.

* Migration tool: issues in FAQ pages - move no-scroll to gatsby-theme-bodiless.

* Migration tool: issues in FAQ pages - refactoring selector.

* Migration tool: issues in FAQ pages - configure update.

* Migration tool: issues in FAQ pages - documentation update.

* Migration tool: issues in FAQ pages - fix edge cases.

* Migration tool: issues in FAQ pages - documentation update.

* Migration tool: issues in FAQ pages - documentation update.

wodenx committed:
fix(psh): Private npm registry for static site. (#148)

## Changes
- Update static.platform.sh to add a .npmrc file specifyijng a private registry for namespaced packages.  Functionality already existed on edit site; this adds it to static.

beliayeu committed:
address pr feedback

wodenx committed:
Direction for state handlign on delete item.

VancheeZze committed:
fix(gatsby-theme-bodiless): Fix revert does not refresh page (#110)

## Changes
- Updated Gatsby version to 2.19.17
- Add page reload after git revert
- Update Revert confirmation message 

## Test Instructions

- Run `npm run start`
- Make changes when in edit mode
- Click the Revert Button
- Confirm Reverting
- All the changes should be reverted.

## Related Issues
Fixes: https://github.com/johnsonandjohnson/Bodiless-JS/issues/109

beliayeu committed:
fix(gatsby-theme-bodiless): Typed chars disappear sporadically (#52)

## Changes
* fix for the problem reported in #14

### Bodiless backend changes
* adjusted bodiless data model. extended json data with metadata. the metadata has author field that contains uuid of a client that made last modification of the data. The client uuid is taken from http request header.

### Bodiless Gatsby theme changes
*  added client uuid to requests sent to backend. The client uuid is generated for MobX store and will change on page reload.
* MobX store rejects data if the data was originated from the same client.

## Test Instructions

see #14

## Related Issues

Fixes #14

beliayeu committed:
fix(gatsby-theme-bodiless): Deleting json file for any component does not reset content (#145)

beliayeu committed:
Merge branch 'master' into feat/retained-data

beliayeu committed:
fix broken tests

wodenx committed:
feat(documentation): Support better control over IA (#129)

* Provide default IA and mechanism to extend it.

* First step towards namespaced doc trees.

* prepending nameSpace works when path is empty.

* WIP - absolute links (currently broken) -- need to copy index.html manually.

* Update path prefix.

* Fix bodiless.docs.json in organisms and accessibility.

* Copy resources; hoist first namespace to root.

* Working pack script but isn't packaged right.

* Working bodiless pack command.

* Fix linting errors.

* Add README to bodiless-cli.

* Fix issue when no devDependencies exist.

* Cleanup of bodiless docs.

* Removed bodiless-cli to another PR

* Update docs and regenerate package-lock.json

* Fix tests

* Fix linting

* Move env var to top level and inject copier to writeTree and writeResources.

Co-authored-by: Erik Summerfield <erik@e2thex.org>
Co-authored-by: hvanyo <58534559+hvanyo@users.noreply.github.com>

beliayeu committed:
chore(bodiless-migration-tool): package-lock.json is updated sporadically with css-to-object changes

dewen committed:
v0.0.40 (#156)

beliayeu committed:
Merge branch 'master' into feat/retained-data

beliayeu committed:
refined the logic when parent list item is deleted

VancheeZze committed:
feat(migration_tool): Migration tool enhancement to auto turn off Bodiless-Tailwind Theme (#155)

## Changes
- New `setupEnvVars` step for the migration tool which currently disables `BODILESS_TAILWIND_THEME_ENABLED` before compilation.
- New optional `disableTailwind ` param for `migration-settings.json` which controles `BODILESS_TAILWIND_THEME_ENABLED`. Set to `true` by default.

## Test Instructions
- Run `npm run sites:clone-local starter`
- Run `npm run sites:update starter`
- Go to `cd sites/starter`
- Make sure `BODILESS_TAILWIND_THEME_ENABLED` is commented out in `.env.site` by default
- Run `npm run migrate`
- Make sure `BODILESS_TAILWIND_THEME_ENABLED` is now set to `0` and not commented out in `.env.site`.

## Related Issues
Closes: https://github.com/johnsonandjohnson/Bodiless-JS/issues/124

dewen committed:
Chore: Example site package dependency update after v0.0.40. (#157)

* chore: example site dependency update after v0.0.40 release.
* Package version update after v0.0.40 release

wodenx committed:
Merge branch 'master' into feat/retained-data

beliayeu committed:
fix: npm run new & npm run sites:update starter fail with errors (#153)

beliayeu committed:
Merge branch 'master' into feat/retained-data

beliayeu committed:
adjust delete sublist mutator

beliayeu committed:
renamed deleteItem into deleteSubnode
